extract_region_feature.py

```diff
@@ -0,0 +1,170 @@
+import numpy as np
+from torch.nn import functional as F
+
+from detectron2.checkpoint import DetectionCheckpointer
+from detectron2.config import get_cfg
+from detectron2.engine import default_setup
+from detectron2.evaluation import inference_context
+from detectron2.modeling import build_model
+from detectron2.modeling import postprocessing
+
+from grid_feats import (
+    add_attribute_config,
+    build_detection_test_loader_with_attributes,
+    build_detection_test_loader_for_images,
+)
+
+# A simple mapper from object detection dataset to VQA dataset names
+dataset_to_folder_mapper = {}
+dataset_to_folder_mapper["coco_2014_train"] = "train2014"
+dataset_to_folder_mapper["coco_2014_val"] = "val2014"
+# One may need to change the Detectron2 code to support coco_2015_test
+# insert "coco_2015_test": ("coco/test2015", "coco/annotations/image_info_test2015.json"),
+# at: https://github.com/facebookresearch/detectron2/blob/master/detectron2/data/datasets/builtin.py#L36
+dataset_to_folder_mapper["coco_2015_test"] = "test2015"
+
+
+def extract_feature_argument_parser():
+    parser = argparse.ArgumentParser(description="Region feature extraction")
+    parser.add_argument(
+        "--config-file", default="", metavar="FILE", help="path to config file"
+    )
+    parser.add_argument(
+        "--dataset",
+        help="name of the dataset",
+        default="coco_2014_train",
+        choices=["coco_2014_train", "coco_2014_val", "coco_2015_test"],
+    )
+    parser.add_argument(
+        "--dataset-path",
+        help="path to image folder dataset, if not provided expects a detection dataset",
+        default="",
+    )
+    parser.add_argument(
+        "--feature-name",
+        help="type of feature to extract, for FPN and DC5 models it can be either of the last two FC regression outputs",
+        default="fc7",
+        choices=["fc6", "fc7"],
+    )
+    parser.add_argument(
+        "opts",
+        help="Modify config options using the command-line",
+        default=None,
+        nargs=argparse.REMAINDER,
+    )
+    return parser
+
+
+def extract_feature_on_dataset(model, data_loader, dump_folder, feature_name):
+    for idx, inputs in enumerate(tqdm.tqdm(data_loader)):
+        with torch.no_grad():
+            image_id = inputs[0]["image_id"]
+
+            # compute features and proposals
+            images = model.preprocess_image(inputs)
+            features = model.backbone(images.tensor)
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            _, pooled_features, pooled_features_fc6 = model.roi_heads.get_roi_features(
+                features, proposals
+            )
+            predictions = model.roi_heads.box_predictor(pooled_features)
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
```

Comment:
Are duplicating r_indices removed in this case? iirc it is removed in the BUTD code. Duplicate means the same proposal box corresponding to two different objects.